### DISCUSSION TITLE
Question about using specific designs in the adjoint optimizer

### DISCUSSION THREAD

**Discussion by pwflanigan** [NanoComp/meep#2759] [Q&A] [answered]
The examples for the adjoint optimizer (https://github.com/NanoComp/meep/tree/master/python/examples/adjoint_optimization) all seem to use either a single fixed value or a random array of 0's and 1's for the initial guess of the design region. Is there a way to feed in a specific design as the initial guess? I just wanted to see if there was an easier built-in mechanism rather than having to do something like:

1. Make a different simulation with the design you want to start with (putting down blocks, cylinders, etc., or importing from GDS)
2. Export that epsilon map to an h5 file
3. Convert the h5 file to a 1-d array with values 0 <= rho <= 1 then setting that as the initial guess

**Comment by smartalecH** [comment]
There's no in-place machinery to make this easier, unfortunately. At least not for now. 

(If you have a gds file, read in the polygons and use a point-in-polygon algorithm to sample the geometry onto the grid. No need to export using a separate simulation, etc)

A tutorial along these lines would be useful. Do you have a design in mind that would be good for a tutorial?

**Comment by pwflanigan** [comment, marked as answer]
I've been playing around with it, and this is the basic idea I've come up with: create a dummy simulation with the objects you want to use as the initial design, convert it to an array with get_epsilon(), then convert that to a [0,1] vector that can be read as the initial design of the optimizer. It gets complicated because the objects seem to scale differently in the design region and the main sim; I'm still trying to get a handle on how it works, but it doesn't like there's any way to have a non-square design region, at least not at this point.

Here is an example, based off of <https://github.com/NanoComp/meep/blob/master/python/examples/adjoint_optimization/01-Introduction.ipynb>:

```
import meep as mp
import meep.adjoint as mpa
import numpy as np
from autograd import numpy as npa
from matplotlib import pyplot as plt
import nlopt

nSi = 3.40
ep_max = nSi**2
Si = mp.Medium(index=nSi)

nSiO2 = 1.44
ep_min = nSiO2**2 
SiO2 = mp.Medium(index=nSiO2)

res_cell = 20
Scellx = 7.0
Scelly = 7.0
cell_size = mp.Vector3(Scellx,Scelly,0)
tpml = 1.0
pml_layers = [mp.PML(tpml)]

fcen = 1/1.55
source = [mp.EigenModeSource(
        mp.GaussianSource(frequency=fcen,fwidth=0.1*fcen),
        eig_band=1,
        direction=mp.NO_DIRECTION,
        eig_kpoint=mp.Vector3(1,0,0),
        size=mp.Vector3(0,2,0),
        center=mp.Vector3(-Scellx/2+tpml,0,0))]

design_region_resolution = 60
Nrho = design_region_resolution + 1
Srhox = 3.0
Srhoy = 3.0

Lwg = (Scellx - Srhox)/2
Wwg = 0.5

design_variables = mp.MaterialGrid(mp.Vector3(Nrho,Nrho,0),SiO2,Si,grid_type="U_DEFAULT") # was U_MEAN
design_region = mpa.DesignRegion(design_variables,volume=mp.Volume(center=mp.Vector3(0,0,0),size=mp.Vector3(Srhox,Srhoy,0)))

geometry = [mp.Block(center=mp.Vector3(-Scellx/2+Lwg/2,0,0),material=Si,size=mp.Vector3(Lwg,Wwg,0)),
    mp.Block(center=mp.Vector3(0,Scelly/2-Lwg/2,0),material=Si,size=mp.Vector3(Wwg,Lwg,0)),
    mp.Block(center=design_region.center,size=design_region.size,material=design_variables)]

sim = mp.Simulation(
    cell_size=cell_size,
    boundary_layers=pml_layers,
    geometry=geometry,
    sources=source,
    eps_averaging=False,
    resolution=res_cell)

TE0 = mpa.EigenmodeCoefficient(
    sim, mp.Volume(center=mp.Vector3(0, 1, 0), size=mp.Vector3(x=2)), mode=1)
ob_list = [TE0]

def J(alpha):
    return npa.abs(alpha) ** 2

opt = mpa.OptimizationProblem(
    simulation=sim,
    objective_functions=J,
    objective_arguments=ob_list,
    design_regions=[design_region],
    fcen=fcen,df=0,nf=1)

DR_cyl_center = mp.Vector3(-0.5,0.5,0) # mp.Vector3(-Srhox/2,Srhoy/2,0)
design_region_geometry = [mp.Cylinder(material=Si,center=DR_cyl_center,radius=0.60),
                          mp.Cylinder(material=mp.air,center=DR_cyl_center,radius=0.40)]

design_region_sim = mp.Simulation(cell_size=mp.Vector3(1,1,0),
                    resolution=Nrho,
                    geometry=design_region_geometry,
                    eps_averaging=False)
                    
plt.figure(dpi=150)
design_region_sim.plot2D()

design_region_array = design_region_sim.get_epsilon()
initial_design = (design_region_array.flatten() - ep_min)/(ep_max - ep_min)

opt.update_design([initial_design])
plt.figure(dpi=150)
opt.plot2D()
```
